beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,705 @@
+        verify_propagation_slot_range(chain, contribution)?;
+
+        // Validate subcommittee index.
+        if contribution.subcommittee_index >= SYNC_COMMITTEE_SUBNET_COUNT {
+            return Err(Error::InvalidSubcommittee {
+                subcommittee_index: contribution.subcommittee_index,
+                subcommittee_size: SYNC_COMMITTEE_SUBNET_COUNT,
+            });
+        }
+
+        // Ensure the valid sync contribution has not already been seen locally.
+        let contribution_root = contribution.tree_hash_root();
+        if chain
+            .observed_sync_contributions
+            .write()
+            .is_known(contribution, contribution_root)
+            .map_err(|e| Error::BeaconChainError(e.into()))?
+        {
+            return Err(Error::SyncContributionAlreadyKnown(contribution_root));
+        }
+
+        // Ensure there has been no other observed aggregate for the given `aggregator_index`.
+        //
+        // Note: do not observe yet, only observe once the sync contribution has been verified.
+        let observed_key =
+            SlotSubcommitteeIndex::new(contribution.slot, contribution.subcommittee_index);
+        match chain
+            .observed_sync_aggregators
+            .read()
+            .validator_has_been_observed(observed_key, aggregator_index as usize)
+        {
+            Ok(true) => Err(Error::AggregatorAlreadyKnown(aggregator_index)),
+            Ok(false) => Ok(()),
+            Err(ObservedAttestersError::ValidatorIndexTooHigh(i)) => {
+                Err(Error::ValidatorIndexTooHigh(i))
+            }
+            Err(e) => Err(BeaconChainError::from(e).into()),
+        }?;
+
+        // Ensure the block being voted for (contribution.beacon_block_root) passes validation.
+        //
+        // This indirectly checks to see if the `contribution.beacon_block_root` is in our fork
+        // choice. Any known, non-finalized, processed block should be in fork choice, so this
+        // check immediately filters out sync committee contributions that attest to a block that
+        // has not been processed.
+        //
+        // Sync committee contributions must be for a known block. If the block is unknown, we
+        // simply drop the sync committee contribution and do not delay consideration for later.
+        verify_head_block_is_known(chain, contribution.beacon_block_root)?;
+
+        // Ensure that the sync committee message has participants.
+        if contribution.aggregation_bits.is_zero() {
+            return Err(Error::EmptyAggregationBitfield);
+        }
+
+        // Ensure the aggregator's pubkey is in the declared subcommittee of the current sync committee
+        let pubkey_bytes = chain
+            .validator_pubkey_bytes(aggregator_index as usize)?
+            .ok_or(Error::UnknownValidatorIndex(aggregator_index as usize))?;
+
+        let current_sync_committee = chain.head_current_sync_committee()?;
```

Comment:
Implemented this but having some trouble figuring out how to best add a test. Since the first sync committee period has `current_sync_committee` equal to `next_sync_committee` seems like we'll need to advance to slot `(512 * 2 * 32) - 1 `. Might try to switch to the minimial config... also experimenting with advancing the test harness faster

Reply:
Ended up just using `harness.add_attested_block_at_slot()` for the slot before the second sync committee boundary. It's causing the tests to run at around 90 seconds for me, but doesn't seem too bad. Haven't found any other good way to speed up the harness.